eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala

```diff
@@ -306,6 +315,21 @@ trait Service extends Logging {
       }
     }
 
+  def makeSocketHandler(system: ActorSystem)(implicit materializer: ActorMaterializer): Flow[Message, TextMessage.Strict, NotUsed] = {
+    val (client, source) = Source.actorRef[String](10, OverflowStrategy.dropTail).toMat(BroadcastHub.sink[String])(Keep.both).run()
+    val producer = source.toMat(BroadcastHub.sink(bufferSize = 256))(Keep.right).run()
+
+    system actorOf Props(new Actor {
+      override def preStart: Unit = context.system.eventStream.subscribe(self, classOf[PaymentReceived])
+      def receive: Receive = { case received: PaymentReceived => client ! received.paymentHash.toString }
+    })
+
+    Flow[Message]
+      .mapConcat(_ => Nil) // Ignore heartbeats and other data from the client
+      .merge(producer)  // Stream the data we want to the client
+      .map(TextMessage.apply)
+  }
+
```

Comment:
Few remarks:
* variable names are not very clear
* this section deserves a few comments because akka stream code is not very readable

Also, `producer` does not seem necessary, I made a few modifications and it works ok (I'm using a `Queue` instead of an `Actor` but the code is the same).

```scala
def makeSocketHandler(system: ActorSystem)(implicit materializer: ActorMaterializer) = {

    // create a flow transforming a queue of string -> string
    val (flowInput, flowOutput) = Source.queue[String](10, OverflowStrategy.dropTail).toMat(BroadcastHub.sink[String])(Keep.both).run()

    // register an actor that feeds the queue when a payment is received
    system.actorOf(Props(new Actor {
      override def preStart: Unit = context.system.eventStream.subscribe(self, classOf[PaymentReceived])
      def receive: Receive = { case received: PaymentReceived => flowInput.offer(received.paymentHash.toString) }
    }))

    Flow[Message]
      .mapConcat(_ => Nil) // Ignore heartbeats and other data from the client
      .merge(flowOutput) // Stream the data we want to the client
      .map(TextMessage.apply)
  }
```

Maybe I am missing something?

Reply:
Seems like nothing is being missed, it works either way. I've taken a more low level approach but looks like there is no point to it.